src/lfortran/semantics/ast_common_visitor.h

```diff
@@ -10166,7 +10166,11 @@ class CommonVisitor : public AST::BaseVisitor<Derived> {
                 }
                 if(ASR::is_a<ASR::Allocatable_t>(*array_type)){
                     ASR::down_cast<ASR::Array_t>((ASR::down_cast<ASR::Allocatable_t>(array_type))->m_type)->m_type = ASRUtils::type_get_past_array(ASRUtils::type_get_past_allocatable(tmp2->m_type));
-                    tmp2->m_type = array_type;
+                    if (ASRUtils::is_allocatable(ASRUtils::EXPR(tmp))) {
+                        tmp2->m_type = array_type;
+                    } else {
+                        tmp2->m_type = ASRUtils::type_get_past_allocatable(array_type);
+                    }
```

Comment:
```suggestion
                   tmp2->m_type = ASRUtils::type_get_past_allocatable(array_type);
```
You don't need the if-else condition I believe. The function type_get_past_allocatable already does that.

Reply:
Actually , array-type initially is allocatable . We need to do `ASRUtils::type_get_past_allocatable(array_type)` to get rid of allocatable only in the case when `ASRUtils::EXPR(tmp)` is not allocatable .

Reply:
Oh I see. Yep, this is good then.